xds/internal/resolver/xds_resolver.go

```diff
@@ -236,94 +290,250 @@ func (r *xdsResolver) sendNewServiceConfig(cs *configSelector) bool {
+		// If there was a previous route config watcher because of a non-inline
+		// route configuration, cancel it.
+		r.rdsResourceName = ""
+		if r.routeConfigWatcher != nil {
+			r.routeConfigWatcher.stop()
+			r.routeConfigWatcher = nil
+		}
+
+		r.applyRouteConfigUpdate(*update.InlineRouteConfig)
+		return
 	}
-	if r.xdsClientClose != nil {
-		r.xdsClientClose()
+
+	// We get here only if there was no inline route configuration.
+
+	// If the route config name has not changed, send an update with existing
+	// route configuration and the newly received listener configuration.
+	if r.rdsResourceName == update.RouteConfigName {
+		r.onResolutionComplete()
+		return
 	}
-	r.closed.Fire()
-	r.logger.Infof("Shutdown")
+
+	// If the route config name has changed, cancel the old watcher and start a
+	// new one. At this point, since we have not yet resolved the new route
+	// config name, we don't have to send an update to the channel.
+	r.rdsResourceName = update.RouteConfigName
+	if r.routeConfigWatcher != nil {
+		r.routeConfigWatcher.stop()
+		r.currentVirtualHost = nil
+		r.routeConfigUpdateRecvd = false
+	}
+	r.routeConfigWatcher = newRouteConfigWatcher(r.rdsResourceName, r)
+}
+
+// onError propagates the error up to the channel. And since this is invoked
+// only for non resource-not-found error, we don't have to update resolver state
+// and we can keep using the old config.
+func (r *xdsResolver) onError(err error) {
+	r.cc.ReportError(err)
+}
+
+// Contains common functionality to be executed when resources of either type
+// are removed.
+func (r *xdsResolver) onResourceNotFound() {
+	// We cannot remove clusters from the service config that have ongoing RPCs.
+	// Instead, what we can do is to send an erroring (nil) config selector
+	// along with normal service config. This will ensure that new RPCs will
+	// fail, and once the active RPCs complete, the reference counts on the
+	// clusters will come down to zero. At that point, we will send an empty
+	// service config config with no addresses. This results in the pick-first
+	// LB policy being configured on the channel, and since there are no
+	// address, pick-first will put the channel in TRANSIENT_FAILURE.
+	r.sendNewServiceConfig(nil)
+
+	// Stop and dereference the active config selector, if one exists.
+	r.curConfigSelector.stop()
+	r.curConfigSelector = nil
+}
+
+func (r *xdsResolver) onListenerResourceNotFound() {
+	if r.routeConfigWatcher != nil {
+		r.routeConfigWatcher.stop()
+	}
+	r.rdsResourceName = ""
+	r.currentVirtualHost = nil
+	r.routeConfigUpdateRecvd = false
+	r.routeConfigWatcher = nil
+
+	r.onResourceNotFound()
+}
+
+func (r *xdsResolver) onRouteConfigResourceUpdate(name string, update xdsresource.RouteConfigUpdate) {
+	if r.rdsResourceName != name {
+		// Drop updates from canceled watchers.
+		return
+	}
+
+	r.applyRouteConfigUpdate(update)
+}
+
+func (r *xdsResolver) onRouteConfigResourceNotFound(name string) {
+	if r.rdsResourceName != name {
+		return
+	}
```

Comment:
Does this not need to do more cleanup? I.e. at least r.receivedRouteConfigUpdate = false (and niling currentVh)? Is resource not found for a route a terminal error? Do we treat the same as LDS on error not found? On server side, resource not found for LDS moves server into non serving and for RDS is treated as routing configuration which fails at the L7 layer (i.e. an RPC comes in), thus the logic is different. Right now both put channel in Transient Failure. Is that what we want?

Reply:
A route configuration once received cannot move to a not-found state. This is by design of LDS/CDS vs RDS/EDS.

An RDS/EDS resource can hit the resource-not-found error *only* in the case where it is never received from the management server and the watch timer expires.

Reply:
I agree, it can only hit because of the 20 second timeout because no diffs. However, my question to you is what should the behavior be in this case. On server side, we fail RPCs at the L7 level, but move the Server to Serving.

Reply:
I.e. should this not do what lis resource not found does and nil currentVh and set receivedRouteConfigUpdate to false? It sends a new erroring config selector like LDS but doesn't clear out it's state (I don't think this affects logic as it's written), but clearing out state feels like the appropriate thing to do here.

Reply:
The state would never have been set in the first place because this error can only happen before we see the first update (which is when `currentVh` and `receivedRouteConfigUpdate` are set).

Reply:
No, it can get a first LDS update which specifies RDS1, then a new LDS update which specifies RDS2, and then the RDS2 can throw a resource not found due to the timer.

Reply:
The state is reset here when that happens: https://github.com/grpc/grpc-go/pull/6729/files#diff-9a978ead61ce685fb95c516a4767ac31d62998e4095eceb516498ac7b34ce16cR506

Reply:
Ah ok that looks good to me